### ISSUE TITLE
tsc does not infer "this" correctly in MyClass.prototype.getData syntax

### ISSUE THREAD

**Issue by michael-brade** [Microsoft/TypeScript#13621]
**TypeScript Version:**  2.1.4

**Code**

Consider this:
```ts
class MyClass {
  data: string;
  wrongData: {};
}

interface MyClass {
  getData(): string;
}

MyClass.prototype.getData = function() {
  return this.wrongData; // this.data;
};
```

**Expected behavior:**

VS Code should show an error, tsc should refuse to compile it. Hovering `this` in VS Code should show `MyClass`, not `any`.

**Actual behavior:**

TypeScript 2.1.4 compiles this without complaining. VS Code doesn't show an error. It infers correctly that getData() should return a string, but it thinks `this` is `any`.

**Comment by mhegazy** [comment]
If `this` type is not specified, it is inferred to be `any`. consider adding a `this` declaration to your interface:
```ts
interface MyClass {
  getData(this:MyClass): string;
}
```

**Comment by michael-brade** [comment]
Well, I guess this is a bit more verbose but it would work as a workaround for now. However, it looks ugly and seems totally unneccessary. MyClass.prototype.xxxx already implies that `this` is `MyClass`, doesn't it? So why can't or shouldn't TypeScript infer this correctly?

**Comment by mhegazy** [comment]
Please see https://github.com/Microsoft/TypeScript/issues/10288
